File: appMain/src/main/java/com/cartlc/tracker/data/DataPicture.java
/**
 * Copyright 2018, FleetTLC. All rights reserved
 */
package com.cartlc.tracker.data;

import android.text.TextUtils;

import com.cartlc.tracker.util.BitmapHelper;

import java.io.File;

import timber.log.Timber;

/**
 * Created by dug on 5/16/17.
 */

public class DataPicture {

    static final int MAX_NOTE_LENGTH = 1000;

    public long    id;
    public String  unscaledFilename;
    public String  scaledFilename;
    public String  note;
    public boolean uploaded;

    File unscaledFile;
    File scaledFile;

    public DataPicture() {
    }

    public DataPicture(long id, String pictureFilename, String uploadingFilename, String note, boolean uploaded) {
        this.id = id;
        this.unscaledFilename = pictureFilename;
        this.scaledFilename = uploadingFilename;
        this.note = note;
        this.uploaded = uploaded;
    }

    public File getUnscaledFile() {
        if (unscaledFile == null) {
            unscaledFile = new File(unscaledFilename);
        }
        return unscaledFile;
    }

    public String getTailname() {
        int pos = unscaledFilename.lastIndexOf("/");
        if (pos >= 0) {
            return unscaledFilename.substring(pos + 1);
        }
        return unscaledFilename;
    }

    public boolean existsUnscaled() {
        return getUnscaledFile().exists();
    }

    public boolean existsScaled() {
        if (getScaledFile() != null) {
            return getScaledFile().exists();
        }
        return false;
    }

    public void remove() {
        getUnscaledFile().delete();
        if (getScaledFile() != null) {
            getScaledFile().delete();
        }
    }

    // Warning: will create the scaled file if it does not yet exist.
    public File getScaledFile() {
        if (scaledFilename == null) {
            scaledFilename = BitmapHelper.createScaled(getUnscaledFile());
            if (scaledFilename != null) {
                TablePictureCollection.getInstance().update(this, null);
            } else {
                Timber.e("Failed to create uploading version of " + getUnscaledFile().getAbsolutePath());
            }
        }
        if (scaledFile == null && scaledFilename != null) {
            scaledFile = new File(scaledFilename);
        }
        return scaledFile;
    }

    public void setNote(String note) {
        // Safety:
        if (note.length() > MAX_NOTE_LENGTH) {
            note = note.substring(0, MAX_NOTE_LENGTH);
        }
        this.note = note;
        TablePictureCollection.getInstance().update(this, null);
    }

    public int rotateCW() {
        BitmapHelper.rotate(getUnscaledFile(), 90);
        return 90;
    }

    public int rotateCCW() {
        BitmapHelper.rotate(getUnscaledFile(), -90);
        return -90;
    }

    public String toString() {
        StringBuilder sbuf = new StringBuilder();
        sbuf.append(id);
        if (unscaledFilename != null) {
            sbuf.append(", ");
            sbuf.append(unscaledFilename);
        }
        if (scaledFilename != null) {
            sbuf.append(", ");
            sbuf.append(scaledFilename);
        }
        if (!TextUtils.isEmpty(note)) {
            sbuf.append(", note=");
            sbuf.append(note);
        }
        if (uploaded) {
            sbuf.append(", UPLOADED");
        }
        return sbuf.toString();
    }
}

File: appMain/src/main/java/com/cartlc/tracker/model/data/DataPicture.kt
/**
 * Copyright 2018, FleetTLC. All rights reserved
 */
package com.cartlc.tracker.model.data

import com.cartlc.tracker.ui.util.helper.BitmapHelper

import java.io.File

/**
 * Created by dug on 5/16/17.
 */

class DataPicture(
        _id: Long = 0,
        val unscaledFilename: String,
        _scaledFilename: String? = null,
        _note: String? = null,
        _uploaded: Boolean = false
) {

    companion object {
        internal val MAX_NOTE_LENGTH = 1000
    }

    var id: Long = _id

    var scaledFilename: String? = _scaledFilename
        get() {
            if (field == null) {
                field = BitmapHelper.createScaledFilename(unscaledFilename)
            }
            return field
        }

    var note: String? = _note
        set(value) {
            if (value == null) {
                field = value
            } else {
                if (value.length > MAX_NOTE_LENGTH) {
                    field = value.substring(0, MAX_NOTE_LENGTH)
                } else {
                    field = value
                }
            }
        }

    var uploaded: Boolean = _uploaded

    val existsScaled: Boolean
        get() = scaledFile?.exists() ?: false

    val unscaledFile: File
        get() = File(unscaledFilename)

    val existsUnscaled: Boolean
        get() = unscaledFile.exists()

    val scaledFile: File?
        get() = scaledFilename?.let { return File(scaledFilename) }

    fun buildScaledFile() = scaledFilename?.let {
        BitmapHelper.createScaled(unscaledFile, it)
    } ?: false

    val tailname: String
        get() {
            val pos = unscaledFilename.lastIndexOf("/")
            return if (pos >= 0) {
                unscaledFilename.substring(pos + 1)
            } else unscaledFilename
        }

    fun remove() {
        unscaledFile.delete()
        scaledFile?.delete()
    }

    fun rotateCW(): Int {
        BitmapHelper.rotate(unscaledFile, 90)
        return 90
    }

    fun rotateCCW(): Int {
        BitmapHelper.rotate(unscaledFile, -90)
        return -90
    }

    override fun toString(): String {
        val sbuf = StringBuilder()
        sbuf.append(id)
        sbuf.append(", ")
        sbuf.append(unscaledFilename)
        if (scaledFilename != null) {
            sbuf.append(", ")
            sbuf.append(scaledFilename)
        }
        if (!note.isNullOrEmpty()) {
            sbuf.append(", note=")
            sbuf.append(note)
        }
        if (uploaded) {
            sbuf.append(", UPLOADED")
        }
        return sbuf.toString()
    }
}
